File: Java/2021/Day03/part2.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.*;

class Part2 {
  public static void main(String[] args) throws FileNotFoundException {
    ArrayList<String> data = readFile("input.txt");

    if (data.size() == 0) {
      return;
    }

    int oxygen = extractData(data, "oxygen");
    int co2 = extractData(data, "co2");

    System.out.println(oxygen*co2);
  }

  private static int extractData(ArrayList<String> data, String type) {
    for(int i = 0; i < data.get(0).length(); i++) {
      int ones = 0;
      int zeros = 0;

      for (int j = 0; j < data.size(); j++) {
        if(data.get(j).charAt(i) == '1') {
          ones++;
        } else {
          zeros++;
        }
      }

      char toKeep;

      if((type == "oxygen" && (ones >= zeros)) || (type == "co2" && (ones < zeros))) {
        toKeep = '1';
      } else {
        toKeep = '0';
      }

      ArrayList<String> dataCopy = new ArrayList<String>();

      for (int j = 0; j < data.size(); j++) {
        if(data.get(j).charAt(i) == toKeep) {
          dataCopy.add(data.get(j));
        }
      }

      data = dataCopy;

      if (data.size() == 1) {
        break;
      }
    }

    return Integer.parseInt(data.get(0), 2);
  }

  private static ArrayList<String> readFile(String path) throws FileNotFoundException {
    File myFile = new File(path);
    Scanner fileScanner = new Scanner(myFile);

    ArrayList<String> data = new ArrayList<String>();

    while (fileScanner.hasNextLine()) {
      String line = fileScanner.nextLine().trim();

      data.add(line);
    }

    return data;
  }
}

File: Python/2021/Day03/part2.py
file = "input.txt"

values = []
with open(file, "r") as open_file:
    while (line := open_file.readline()):
        values.append(
            line.strip()
        )

valuesCopy = values.copy()

oxygen = 0

for i in range(len(values[0])):
    ones = 0
    zeros = 0

    for j in range(len(values)):
        if values[j][i] == "1":
            ones += 1
        else:
            zeros += 1

    if ones >= zeros:
        keepWith = "1"
    else:
        keepWith = "0"

    for value in values.copy():
        if value[i] != keepWith:
            values.remove(value)

    if len(values) == 1:
        break

oxygen = int(values[0],2)

values = valuesCopy.copy()

co2 = 0
for i in range(len(values[0])):
    ones = 0
    zeros = 0

    for j in range(len(values)):
        if values[j][i] == "1":
            ones += 1
        else:
            zeros += 1

    if ones < zeros:
        keepWith = "1"
    else:
        keepWith = "0"

    for value in values.copy():
        if value[i] != keepWith:
            values.remove(value)

    if len(values) == 1:
        break

co2 = int(values[0], 2)
print(oxygen*co2)
